python/cudf/cudf/core/column/categorical.py

```diff
@@ -496,56 +570,55 @@ def as_datetime_column(self, dtype, **kwargs):
 
     def _get_decategorized_column(self):
         if self.null_count == len(self):
-            # self._categories is empty; just return codes
+            # self.categories is empty; just return codes
             return self.cat().codes._column
-        gather_map = (
-            self.cat().codes.astype("int32").fillna(0)._column.data.mem
-        )
-        out = self._categories.take(gather_map)
-        return out.replace(mask=self.mask)
+        gather_map = self.cat().codes.astype("int32").fillna(0)._column
+        out = self.categories.take(gather_map)
+        out.mask = self.mask
+        return out
 
     def copy(self, deep=True):
-        """Categorical Columns are immutable, so a deep copy produces a
-        copy of the underlying data, mask, categories and a shallow copy
-        creates a new column and copies the references of the data, mask
-        and categories.
-        """
         if deep:
             copied_col = libcudf.copying.copy_column(self)
-            category_col = libcudf.copying.copy_column(self._categories)
-            return self.replace(
-                data=copied_col.data,
-                mask=copied_col.mask,
+            return column.build_column(
+                data=None,
                 dtype=self.dtype,
-                categories=category_col,
-                ordered=self._ordered,
+                mask=copied_col.mask,
+                name=self.name,
+                children=(copied_col,),
             )
         else:
-            params = self._replace_defaults()
-            return type(self)(**params)
+            return column.build_column(
+                data=None,
+                dtype=self.dtype,
+                mask=self.mask,
+                name=self.name,
+                children=self.children,
+            )
 
 
 def pandas_categorical_as_column(categorical, codes=None):
+
     """Creates a CategoricalColumn from a pandas.Categorical
 
     If ``codes`` is defined, use it instead of ``categorical.codes``
     """
     # TODO fix mutability issue in numba to avoid the .copy()
     codes = categorical.codes.copy() if codes is None else codes
+    codes = column.as_column(codes)
     # TODO pending pandas to be improved
```

Comment:
This comment can be removed now

Reply:
Which one? Wasn't sure

Reply:
I believe we can remove both and remove the `.copy()` call above. If that gives issues then just the Pandas improvement comment.

Reply:
Done